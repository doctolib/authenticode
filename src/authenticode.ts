import { promisify } from 'util'
import { execFile } from 'child_process'

// https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.signaturestatus
export enum SignatureStatus {
  Valid = 0,
  UnknownError = 1,
  NotSigned = 2,
  HashMismatch = 3,
  NotTrusted = 4,
  NotSupportedFileFormat = 5,
  Incompatible = 6,
}

export enum SignatureType {
  None = 0,
  Authenticode = 1,
  Catalog = 2,
}

// https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2?redirectedfrom=MSDN&view=netcore-3.1
export interface SignerCertificate {
  Subject: string
}

// https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.signature
export interface BaseSignature {
  IsOSBinary: unknown
  Path: string
  SignatureType: SignatureType
  SignerCertificate: unknown
  Status: SignatureStatus
  StatusMessage: string
  TimeStamperCertificate: unknown
}

const statusWithSignature = [SignatureStatus.Valid, SignatureStatus.HashMismatch, SignatureStatus.NotTrusted] as const

export interface PresentSignature extends BaseSignature {
  Status: typeof statusWithSignature[number]
  SignerCertificate: SignerCertificate
}

export type Signature = BaseSignature | PresentSignature

export interface SignatureSummary {
  status: string
  type: string
  subject: string | null
}

export async function getAuthenticode(path: string): Promise<Signature> {
  const { stdout } = await promisify(execFile)('powershell.exe', [
    '-NoProfile',
    '-Command',
    "'Get-AuthenticodeSignature $args | ConvertTo-Json -Compress'",
    path,
  ])
  return JSON.parse(stdout)
}

function signatureIsPresent(signature: Signature): signature is PresentSignature {
  return (statusWithSignature as readonly SignatureStatus[]).includes(signature.Status)
}

export function summary(signature: Signature): SignatureSummary {
  return {
    status: SignatureStatus[signature.Status] ?? String(signature.Status),
    type: SignatureType[signature.SignatureType] ?? String(signature.SignatureType),
    subject: signatureIsPresent(signature) ? signature.SignerCertificate.Subject : null,
  }
}
