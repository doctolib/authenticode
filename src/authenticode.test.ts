import { getAuthenticode, summary } from './authenticode'

function path(kind: string): string {
  return `./src/sign-${kind}.ps1`
}

describe(getAuthenticode, () => {
  it.each([
    ['good', expect.anything()],
    ['bad', expect.anything()],
    ['ugly', null],
  ])('result matches snapshot for %s signature', async (kind, SignerCertificate) => {
    const signature = await getAuthenticode(path(kind))
    expect(signature).toMatchSnapshot({
      StatusMessage: expect.any(String),
      Path: expect.any(String),
      SignerCertificate,
    })
  })
})

const doctolibSubject = [
  'E=connectors@doctolib.com',
  'CN=Doctolib SAS',
  'O=Doctolib SAS',
  'STREET=32 Rue de Monceau',
  'L=Paris',
  'S=Ile de France',
  'C=FR',
  'OID.1.3.6.1.4.1.311.60.2.1.3=FR',
  'SERIALNUMBER=794598813',
  'OID.2.5.4.15=Private Organization',
].join(', ')

describe(summary, () => {
  it.each([
    ['good', { status: 'Valid', type: 'Authenticode', subject: doctolibSubject }],
    ['bad', { status: 'HashMismatch', type: 'Authenticode', subject: doctolibSubject }],
    ['ugly', { status: 'NotSigned', type: 'None', subject: null }],
  ])('result matches snapshot for %s signature', async (kind, expected) => {
    const signature = await getAuthenticode(path(kind))
    expect(summary(signature)).toEqual(expected)
  })
})
