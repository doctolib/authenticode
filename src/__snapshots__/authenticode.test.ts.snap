// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAuthenticode result matches snapshot for bad signature 1`] = `
Object {
  "IsOSBinary": false,
  "Path": Any<String>,
  "SignatureType": 1,
  "SignerCertificate": Anything,
  "Status": 3,
  "StatusMessage": Any<String>,
  "TimeStamperCertificate": null,
}
`;

exports[`getAuthenticode result matches snapshot for good signature 1`] = `
Object {
  "IsOSBinary": false,
  "Path": Any<String>,
  "SignatureType": 1,
  "SignerCertificate": Anything,
  "Status": 0,
  "StatusMessage": Any<String>,
  "TimeStamperCertificate": null,
}
`;

exports[`getAuthenticode result matches snapshot for ugly signature 1`] = `
Object {
  "IsOSBinary": false,
  "Path": Any<String>,
  "SignatureType": 0,
  "SignerCertificate": null,
  "Status": 2,
  "StatusMessage": Any<String>,
  "TimeStamperCertificate": null,
}
`;
